File: backend/engine/scheduler.py
"""Core scheduling engine — turns tasks into a daily timetable."""

from datetime import date, time, datetime, timezone
from typing import List, Tuple, Optional
from dataclasses import dataclass
from backend.config import config
from backend.engine.time_utils import (
    parse_time_str, time_to_minutes, minutes_to_time,
    find_free_slots, fits_in, slot_duration_minutes,
)


@dataclass
class TaskItem:
    """Lightweight task representation for the scheduler (doesn't need DB)."""
    id: int
    title: str
    task_type: str        # fixed | flexible | protected | emergent
    priority: str         # urgent | normal | low
    estimated_minutes: int
    time_hint: Optional[str] = None   # "09:00" or None for flexible
    deadline: Optional[datetime] = None


@dataclass
class SlotOutput:
    """Output slot from the scheduler."""
    task_id: int
    label: str
    start_time: str       # "HH:MM"
    end_time: str         # "HH:MM"
    color: str
    is_locked: bool


@dataclass
class ScheduleResult:
    """Complete schedule result."""
    date: date
    slots: List[SlotOutput]
    deferred: List[TaskItem]   # tasks that didn't fit
    message: str
    tips: List[str] = None     # caring coach suggestions


def _priority_score(priority: str) -> int:
    """Higher score → schedule first."""
    return {"urgent": 100, "normal": 50, "low": 0}.get(priority, 50)


def _color_for_type(task_type: str) -> str:
    return {
        "fixed": "#EF4444",      # red
        "protected": "#10B981",  # green
        "flexible": "#3B82F6",   # blue
        "emergent": "#F59E0B",   # amber
    }.get(task_type, "#6B7280")


def schedule_day(
    schedule_date: date,
    fixed_tasks: List[TaskItem],
    flexible_tasks: List[TaskItem],
    protection_rules: dict,
    emergent_tasks: Optional[List[TaskItem]] = None,
    day_start: time = None,
    day_end: time = None,
    schedule_policy: str = "efficiency_first",
) -> ScheduleResult:
    """
    Build a daily schedule.

    Args:
        schedule_date: The date to schedule for
        fixed_tasks: Fixed-time tasks (courses, meetings) - must have time_hint
        flexible_tasks: Flexible tasks to fill gaps
        protection_rules: {"breakfast": {"start":"07:00","end":"08:00"}, ...}
        emergent_tasks: New emergent tasks to integrate
        day_start: Earliest schedulable time (default 07:00)
        day_end: Latest schedulable time (default 23:00)
        schedule_policy: hard_time_first | deadline_first | efficiency_first

    Returns:
        ScheduleResult with slots, deferred tasks, and a message
    """
    if day_start is None:
        day_start = parse_time_str(config.day_start)
    if day_end is None:
        day_end = parse_time_str(config.day_end)

    all_slots: List[SlotOutput] = []
    occupied: List[Tuple[time, time]] = []

    # -- Step 1: Place fixed tasks --
    for task in fixed_tasks:
        if not task.time_hint:
            continue
        try:
            start = parse_time_str(task.time_hint)
            end = minutes_to_time(time_to_minutes(start) + task.estimated_minutes)
        except (ValueError, AttributeError):
            continue
        all_slots.append(SlotOutput(
            task_id=task.id, label=task.title,
            start_time=f"{start.hour:02d}:{start.minute:02d}",
            end_time=f"{end.hour:02d}:{end.minute:02d}",
            color=_color_for_type(task.task_type), is_locked=True,
        ))
        occupied.append((start, end))

    # -- Step 2: Apply protection rules --
    for rule in protection_rules.values():
        try:
            ps = parse_time_str(rule["start"])
            pe = parse_time_str(rule["end"])
            # Handle overnight (sleep 23:00-07:00)
            if time_to_minutes(pe) <= time_to_minutes(ps):
                occupied.append((ps, time(23, 59)))
                occupied.append((time(0, 0), pe))
            else:
                occupied.append((ps, pe))
        except (KeyError, ValueError):
            continue

    # Add explicit protected tasks from DB
    for task in flexible_tasks:
        if task.task_type == "protected" and task.time_hint:
            try:
                ps = parse_time_str(task.time_hint)
                pe = minutes_to_time(time_to_minutes(ps) + task.estimated_minutes)
                all_slots.append(SlotOutput(
                    task_id=task.id, label=task.title,
                    start_time=f"{ps.hour:02d}:{ps.minute:02d}",
                    end_time=f"{pe.hour:02d}:{pe.minute:02d}",
                    color=_color_for_type("protected"), is_locked=True,
                ))
                occupied.append((ps, pe))
            except (ValueError, AttributeError):
                continue

    # -- Step 3: Find free slots --
    free_slots = find_free_slots(occupied, day_start, day_end)

    # -- Step 4: Sort flexible tasks by priority --
    # Tasks with a time hint MUST be placed before hint-less tasks,
    # otherwise hint-less tasks fill the gap and block the hinted time.
    flex_pool = [t for t in flexible_tasks if t.task_type == "flexible"]
    if schedule_policy == "hard_time_first":
        sorted_flex = sorted(
            flex_pool,
            key=lambda t: (0 if t.time_hint else 1, -_priority_score(t.priority), t.estimated_minutes)
        )
    elif schedule_policy == "deadline_first":
        sorted_flex = sorted(
            flex_pool,
            key=lambda t: (
                0 if t.time_hint else 1,
                t.deadline or datetime(2099, 1, 1, tzinfo=timezone.utc),
                -_priority_score(t.priority),
            )
        )
    else:  # efficiency_first
        sorted_flex = sorted(
            flex_pool,
            key=lambda t: (0 if t.time_hint else 1, -_priority_score(t.priority), t.estimated_minutes)
        )

    # -- Step 5: Greedy bin-packing --
    deferred: List[TaskItem] = []

    for task in sorted_flex:
        placed = False
        # Try time hint first
        if task.time_hint:
            try:
                hint_start = parse_time_str(task.time_hint)
                hint_end = minutes_to_time(time_to_minutes(hint_start) + task.estimated_minutes)
                for fs in free_slots:
                    if (time_to_minutes(hint_start) >= time_to_minutes(fs[0])
                            and time_to_minutes(hint_end) <= time_to_minutes(fs[1])):
                        all_slots.append(SlotOutput(
                            task_id=task.id, label=task.title,
                            start_time=f"{hint_start.hour:02d}:{hint_start.minute:02d}",
                            end_time=f"{hint_end.hour:02d}:{hint_end.minute:02d}",
                            color=_color_for_type("flexible"), is_locked=False,
                        ))
                        occupied.append((hint_start, hint_end))
                        free_slots = find_free_slots(occupied, day_start, day_end)
                        placed = True
                        break
            except (ValueError, AttributeError):
                pass

        if not placed:
            # Find first free slot that fits
            for fs in free_slots:
                if fits_in(fs, task.estimated_minutes):
                    start = fs[0]
                    end = minutes_to_time(time_to_minutes(start) + task.estimated_minutes)
                    all_slots.append(SlotOutput(
                        task_id=task.id, label=task.title,
                        start_time=f"{start.hour:02d}:{start.minute:02d}",
                        end_time=f"{end.hour:02d}:{end.minute:02d}",
                        color=_color_for_type("flexible"), is_locked=False,
                    ))
                    occupied.append((start, end))
                    free_slots = find_free_slots(occupied, day_start, day_end)
                    placed = True
                    break

        if not placed:
            deferred.append(task)

    # -- Step 6: Integrate emergent tasks --
    if emergent_tasks:
        for task in emergent_tasks:
            placed = False
            if task.time_hint:
                try:
                    estart = parse_time_str(task.time_hint)
                    eend = minutes_to_time(time_to_minutes(estart) + task.estimated_minutes)
                    all_slots.append(SlotOutput(
                        task_id=task.id, label=f"⚠ {task.title}",
                        start_time=f"{estart.hour:02d}:{estart.minute:02d}",
                        end_time=f"{eend.hour:02d}:{eend.minute:02d}",
                        color=_color_for_type("emergent"), is_locked=False,
                    ))
                    placed = True
                except (ValueError, AttributeError):
                    pass
            if not placed:
                free_slots = find_free_slots(occupied, day_start, day_end)
                for fs in free_slots:
                    if fits_in(fs, task.estimated_minutes):
                        start = fs[0]
                        end = minutes_to_time(time_to_minutes(start) + task.estimated_minutes)
                        all_slots.append(SlotOutput(
                            task_id=task.id, label=f"⚠ {task.title}",
                            start_time=f"{start.hour:02d}:{start.minute:02d}",
                            end_time=f"{end.hour:02d}:{end.minute:02d}",
                            color=_color_for_type("emergent"), is_locked=False,
                        ))
                        occupied.append((start, end))
                        placed = True
                        break
            if not placed:
                deferred.append(task)

    # Sort all slots by time
    all_slots.sort(key=lambda s: time_to_minutes(parse_time_str(s.start_time)))

    # Build message
    msg_parts = [f"📅 {schedule_date} 日程已生成"]
    msg_parts.append(f"共 {len(all_slots)} 个时段")
    if deferred:
        names = ", ".join(t.title for t in deferred)
        msg_parts.append(f"⚠ {len(deferred)} 项溢出，建议顺延: {names}")

    return ScheduleResult(
        date=schedule_date,
        slots=all_slots,
        deferred=deferred,
        message=" | ".join(msg_parts),
        tips=_build_tips(all_slots, len(deferred)),
    )


def _build_tips(slots: List[SlotOutput], deferred_count: int) -> List[str]:
    """Generate caring, human coach suggestions from the final schedule."""
    tips: List[str] = []
    if not slots:
        return tips
    try:
        first_start = min(s.start_time for s in slots)
        last_end = max(s.end_time for s in slots)
    except (KeyError, ValueError):
        return tips

    locked = [s for s in slots if s.is_locked]
    if locked:
        tips.append(f"{len(locked)} 项固定任务已锁定，记得提前备好资料，到点直接进入状态 🔒")

    first_min = time_to_minutes(parse_time_str(first_start))
    last_min = time_to_minutes(parse_time_str(last_end))
    if first_min <= 8 * 60:
        tips.append("第一件事开始得早，今晚早点休息，明早才有精神 ☀️")
    if last_min >= 21 * 60:
        tips.append("晚上排到比较晚，睡前留点放松时间，少刷手机哦 🌙")
    if len(slots) >= 8:
        tips.append("今天挺满的，记得留点喝水和起身活动的空档 🍵")
    if deferred_count:
        tips.append(f"有 {deferred_count} 件事排不下，明天优先处理，别焦虑 💛")
    if not tips:
        tips.append("节奏很舒服，记得按计划走，也别给自己太大压力 ✨")
    return tips

File: frontend/src/engine/scheduler.ts
// Core scheduling engine — turns tasks into a daily timetable.
// Ported from backend/engine/scheduler.py + time_utils.py (runs fully offline).

export interface TaskItem {
  id: number
  title: string
  task_type: string // fixed | flexible | protected | emergent
  priority: string // urgent | normal | low
  estimated_minutes: number
  time_hint?: string | null // "09:00" or null
  deadline?: string | null
}

export interface SlotOutput {
  task_id: number
  label: string
  start_time: string // "HH:MM"
  end_time: string // "HH:MM"
  color: string
  is_locked: boolean
}

export interface ScheduleResult {
  date: string
  slots: SlotOutput[]
  deferred: TaskItem[]
  message: string
  tips: string[]
}

export function parseTimeStr(t: string): number {
  const parts = t.trim().split(':')
  return (parseInt(parts[0], 10) || 0) * 60 + (parseInt(parts[1], 10) || 0)
}

export function minutesToTime(m: number): string {
  m = ((m % 1440) + 1440) % 1440
  const h = Math.floor(m / 60)
  const min = m % 60
  return `${String(h).padStart(2, '0')}:${String(min).padStart(2, '0')}`
}

function priorityScore(priority: string): number {
  return { urgent: 100, normal: 50, low: 0 }[priority] ?? 50
}

function colorForType(taskType: string): string {
  return {
    fixed: '#EF4444',
    protected: '#10B981',
    flexible: '#3B82F6',
    emergent: '#F59E0B',
  }[taskType] ?? '#6B7280'
}

function mergeAdjacent(slots: Array<[number, number]>): Array<[number, number]> {
  if (!slots.length) return []
  const sorted = [...slots].sort((a, b) => a[0] - b[0])
  const merged: Array<[number, number]> = [sorted[0]]
  for (let i = 1; i < sorted.length; i++) {
    const [start, end] = sorted[i]
    const last = merged[merged.length - 1]
    if (start <= last[1]) {
      last[1] = Math.max(last[1], end)
    } else {
      merged.push([start, end])
    }
  }
  return merged
}

function findFreeSlots(
  occupied: Array<[number, number]>,
  dayStart: number,
  dayEnd: number,
): Array<[number, number]> {
  const merged = mergeAdjacent(occupied)
  const free: Array<[number, number]> = []
  let cursor = dayStart
  for (const [os, oe] of merged) {
    if (os > cursor) free.push([cursor, os])
    if (oe > cursor) cursor = oe
  }
  if (cursor < dayEnd) free.push([cursor, dayEnd])
  return free
}

function fitsIn(slot: [number, number], durationMinutes: number): boolean {
  return slot[1] - slot[0] >= durationMinutes
}

export interface ScheduleDayOptions {
  schedule_date: string
  fixed_tasks: TaskItem[]
  flexible_tasks: TaskItem[]
  protection_rules: Record<string, { start: string; end: string }>
  emergent_tasks?: TaskItem[]
  day_start?: string
  day_end?: string
  schedule_policy?: string
}

export function scheduleDay(opts: ScheduleDayOptions): ScheduleResult {
  const dayStart = parseTimeStr(opts.day_start ?? '07:00')
  const dayEnd = parseTimeStr(opts.day_end ?? '23:00')

  const allSlots: SlotOutput[] = []
  const occupied: Array<[number, number]> = []

  // Step 1: Place fixed tasks
  for (const task of opts.fixed_tasks) {
    if (!task.time_hint) continue
    const start = parseTimeStr(task.time_hint)
    const end = start + task.estimated_minutes
    allSlots.push({
      task_id: task.id, label: task.title,
      start_time: minutesToTime(start), end_time: minutesToTime(end),
      color: colorForType(task.task_type), is_locked: true,
    })
    occupied.push([start, end])
  }

  // Step 2: Apply protection rules
  for (const rule of Object.values(opts.protection_rules ?? {})) {
    const ps = parseTimeStr(rule.start)
    const pe = parseTimeStr(rule.end)
    if (pe <= ps) {
      occupied.push([ps, 23 * 60 + 59])
      occupied.push([0, pe])
    } else {
      occupied.push([ps, pe])
    }
  }

  // Explicit protected tasks from DB
  for (const task of opts.flexible_tasks) {
    if (task.task_type === 'protected' && task.time_hint) {
      const ps = parseTimeStr(task.time_hint)
      const pe = ps + task.estimated_minutes
      allSlots.push({
        task_id: task.id, label: task.title,
        start_time: minutesToTime(ps), end_time: minutesToTime(pe),
        color: colorForType('protected'), is_locked: true,
      })
      occupied.push([ps, pe])
    }
  }

  let freeSlots = findFreeSlots(occupied, dayStart, dayEnd)

  // Step 4: Sort flexible tasks — time-hinted ones MUST come first
  const flexPool = opts.flexible_tasks.filter((t) => t.task_type === 'flexible')
  const sortedFlex = [...flexPool].sort((a, b) => {
    const ah = a.time_hint ? 0 : 1
    const bh = b.time_hint ? 0 : 1
    if (ah !== bh) return ah - bh
    const ap = -priorityScore(a.priority)
    const bp = -priorityScore(b.priority)
    if (ap !== bp) return ap - bp
    return a.estimated_minutes - b.estimated_minutes
  })

  // Step 5: Greedy bin-packing
  const deferred: TaskItem[] = []

  for (const task of sortedFlex) {
    let placed = false
    if (task.time_hint) {
      const hintStart = parseTimeStr(task.time_hint)
      const hintEnd = hintStart + task.estimated_minutes
      for (const fs of freeSlots) {
        if (hintStart >= fs[0] && hintEnd <= fs[1]) {
          allSlots.push({
            task_id: task.id, label: task.title,
            start_time: minutesToTime(hintStart), end_time: minutesToTime(hintEnd),
            color: colorForType('flexible'), is_locked: false,
          })
          occupied.push([hintStart, hintEnd])
          freeSlots = findFreeSlots(occupied, dayStart, dayEnd)
          placed = true
          break
        }
      }
    }
    if (!placed) {
      for (const fs of freeSlots) {
        if (fitsIn(fs, task.estimated_minutes)) {
          const start = fs[0]
          const end = start + task.estimated_minutes
          allSlots.push({
            task_id: task.id, label: task.title,
            start_time: minutesToTime(start), end_time: minutesToTime(end),
            color: colorForType('flexible'), is_locked: false,
          })
          occupied.push([start, end])
          freeSlots = findFreeSlots(occupied, dayStart, dayEnd)
          placed = true
          break
        }
      }
    }
    if (!placed) deferred.push(task)
  }

  // Step 6: Integrate emergent tasks
  for (const task of opts.emergent_tasks ?? []) {
    let placed = false
    if (task.time_hint) {
      const es = parseTimeStr(task.time_hint)
      const ee = es + task.estimated_minutes
      allSlots.push({
        task_id: task.id, label: `⚠ ${task.title}`,
        start_time: minutesToTime(es), end_time: minutesToTime(ee),
        color: colorForType('emergent'), is_locked: false,
      })
      placed = true
    }
    if (!placed) {
      const fs2 = findFreeSlots(occupied, dayStart, dayEnd)
      for (const fs of fs2) {
        if (fitsIn(fs, task.estimated_minutes)) {
          const start = fs[0]
          const end = start + task.estimated_minutes
          allSlots.push({
            task_id: task.id, label: `⚠ ${task.title}`,
            start_time: minutesToTime(start), end_time: minutesToTime(end),
            color: colorForType('emergent'), is_locked: false,
          })
          occupied.push([start, end])
          placed = true
          break
        }
      }
    }
    if (!placed) deferred.push(task)
  }

  allSlots.sort((a, b) => parseTimeStr(a.start_time) - parseTimeStr(b.start_time))

  const msgParts = [`📅 ${opts.schedule_date} 日程已生成`, `共 ${allSlots.length} 个时段`]
  if (deferred.length) {
    msgParts.push(`⚠ ${deferred.length} 项溢出，建议顺延: ${deferred.map((t) => t.title).join(', ')}`)
  }

  return {
    date: opts.schedule_date,
    slots: allSlots,
    deferred,
    message: msgParts.join(' | '),
    tips: buildTips(allSlots, deferred.length),
  }
}

function buildTips(slots: SlotOutput[], deferredCount: number): string[] {
  const tips: string[] = []
  if (!slots.length) return tips

  let firstStart = slots[0].start_time
  let lastEnd = slots[0].end_time
  for (const s of slots) {
    if (parseTimeStr(s.start_time) < parseTimeStr(firstStart)) firstStart = s.start_time
    if (parseTimeStr(s.end_time) > parseTimeStr(lastEnd)) lastEnd = s.end_time
  }

  const locked = slots.filter((s) => s.is_locked)
  if (locked.length) tips.push(`${locked.length} 项固定任务已锁定，记得提前备好资料，到点直接进入状态 🔒`)

  const firstMin = parseTimeStr(firstStart)
  const lastMin = parseTimeStr(lastEnd)
  if (firstMin <= 8 * 60) tips.push('第一件事开始得早，今晚早点休息，明早才有精神 ☀️')
  if (lastMin >= 21 * 60) tips.push('晚上排到比较晚，睡前留点放松时间，少刷手机哦 🌙')
  if (slots.length >= 8) tips.push('今天挺满的，记得留点喝水和起身活动的空档 🍵')
  if (deferredCount) tips.push(`有 ${deferredCount} 件事排不下，明天优先处理，别焦虑 💛`)
  if (!tips.length) tips.push('节奏很舒服，记得按计划走，也别给自己太大压力 ✨')
  return tips
}
